### DISCUSSION TITLE
How to share the same file sink between different loggers (cateogories)

### DISCUSSION THREAD

**Discussion by gianmarcoodorizzi** [gabime/spdlog#2377] [Q&A] [answered]
Possibly related to [this](https://github.com/gabime/spdlog/discussions/2151), but I am not sure I grasped whether this is possible or it is not.

I know that the idea is to [share a common sink](https://github.com/gabime/spdlog/wiki/2.-Creating-loggers#creating-multiple-file-loggers-with-the-same-output-file) between two loggers, but how can I  share that sink between different classes? I would like to avoid having to pass it to every class (e.g.: in its constructor).

The goal is to have different classes writing to the same log file, but with different categories, for instance:

> [2022-05-13 13:54:07.200] [ClassA] [info] Test entry
> [2022-05-13 13:54:07.200] [ClassB] [info] Test entry
> [2022-05-13 13:54:07.200] [ClassC] [info] Test entry

Any suggestion would be greatly appreciated, thank you!

**Comment by tt4g** [comment, marked as answer]
You can get sink from logger.

```cpp
// get another_logger from registry.
auto another_logger = spdlog::get("another");

// get sink by `spdlog::logger::sinks()`.
auto sinks = another_logger->sinks():
// create new_logger.
auto new_logger = std::make_shared<spdlog::logger>("new_logger", sinks.begin(), sinks.end());

// clone logger (share sinks).
auto clone_logger = another_logger->clone("clone_logger"):
```
Wiki: https://github.com/gabime/spdlog/wiki/4.-Sinks#adding-sinks-to-the-logger-after-creation

**Reply by gianmarcoodorizzi** [reply]
Thank you for this. Would that `new_logger` be thread safe though?

In _main.cpp_:
`std::shared_ptr<spdlog::logger> logger = spdlog::rotating_logger_mt("Main", LOGFILE_PATH, 1024 * 1024 * 5, 3);`

In _MyClass.cpp_:
```
auto sinks = spdlog::get("Main")->sinks();
mLogger = std::make_shared<spdlog::logger>("MyClass", sinks.begin(), sinks.end());
```

**Reply by tt4g** [reply]
As noted in the Wiki (https://github.com/gabime/spdlog/wiki/4.-Sinks#adding-sinks-to-the-logger-after-creation), there is no thread-safe guarantee.

> There is no mutex protecting the sinks vector (which is obvious due to the performance impact) so it's NOT thread-safe!

This is because `spdlog::logger::sinks()` returns `std::vector<std::shared_ptr<spdlog::sinks::sink>> &`.
However, if you do not change the vector returned by `spdlog::logger::sinks()`, it is effectively thread-safe.

**Reply by gianmarcoodorizzi** [reply]
Yes, I am aware of that, but I was referring to the use of the logger itself. Would it be fine to call `mLogger->info("Something")` from different threads?

**Reply by tt4g** [reply]
The logger itself is thread-safe.

**Reply by gianmarcoodorizzi** [reply]
@tt4g: I am trying to flush on info level, so I call:
`spdlog::flush_on(spdlog::level::info);`
in my main file.
However, it seems like the other loggers using the same sink do not properly flush. How can I avoid having to set _flush_on_ policy for each _new_logger_? Thank you!

**Reply by tt4g** [reply]
@gianmarcoodorizzi `spdlog::flush_on()` changes the flush policy for loggers registered in spdlog's global registry.
If you created the logger with `std::make_shared<spdlog::logger>()`, register the logger in the registry with `spdlog::register_logger(std::shared_ptr<spdlog::logger>)`.

If the logger is created by a factory function provided by spdlog (e.g. `spdlog::basic_logger_mt()`), the logger is automatically registered in the registry.

**Reply by gianmarcoodorizzi** [reply]
Thanks! Do I have to specify _flush_on_ policy after registering every new logger to the global registry? It seems like if I do not, log entries are not properly flushed.

**Reply by tt4g** [reply]
You can call `spdlog::initizlie_logger()` instead of `spdlog::register_logger()` to apply the flush policy you previously set with `spdlog::flush_on()`.
Note `spdlog::initizlie_logger()` also applies spdlog global settings other than flash policy to the logger.

**Reply by gianmarcoodorizzi** [reply]
Thank you! I have one more question, though. I create multiple object instances for some of my classes, and registering (or initializing) the log with the same name throws an exception. I did not have that issue before, since I was not registering my logs to the global registry. Is there a way to work around this, i.e.: a way to register loggers with the same name? Thanks.

**Reply by tt4g** [reply]
No. Since spdlog identifies loggers by name and registers them in the global registry, loggers with the same name cannot be registered.

**Reply by gianmarcoodorizzi** [reply]
I see, thanks!
